src/store/token/actions.js

```diff
@@ -203,29 +204,86 @@ export function requestUBIPayout(payout) {
   };
 }
 
-async function loopTransfer(from, to, value, paymentNote) {
-  return await waitAndRetryOnFail(
-    () => {
-      return core.token.transfer(from, to, value, paymentNote);
-    },
-    () => {
-      return true;
-    },
-    {},
-    () => {
-      return core.token.updateTransferSteps(from, to, value);
-    },
-  );
+// Recursive helper function for transfer
+async function loopTransfer(
+  from,
+  to,
+  value,
+  paymentNote,
+  hops,
+  attemptsLeft,
+  errorsMessages = '',
+) {
+  if (attemptsLeft === 0 || hops === 0) {
+    // ran out of attempts or hops, cannot attempt further transfers
+    throw new TransferError(errorsMessages);
+  }
+  try {
+    return await core.token.transfer(from, to, value, paymentNote, hops);
+  } catch (error) {
+    // RETRY when path is not found or is too long
+    // UPDATE edges database when the path fails and then retry
+    // GIVE UP for other errors
+    // ---
+    // no path or path too long
+    if (
+      error.name === 'TransferError' &&
+      (error.code === ErrorCodes.TOO_COMPLEX_TRANSFER || // too many steps in found path
+        error.code === ErrorCodes.UNKNOWN_ERROR || // includes timeout error from api
+        error.code === ErrorCodes.INVALID_TRANSFER) // other errors from find transitive transfer
+    ) {
+      // retry with fewer hops
+      return await loopTransfer(
+        from,
+        to,
+        value,
+        paymentNote,
+        hops - 1,
+        attemptsLeft - 1,
+        errorsMessages.concat(' ', error.message),
+      );
+    }
+    // if the path is found but it is invalid
+    else if (
+      (error.name === 'TransferError' && ErrorCodes.TRANSFER_NOT_FOUND) || // search complete and no path found
+      error.name !== 'TransferError' // includes errors from attempting transfer with an invalid path
+    ) {
+      // update the edges db for trust-adjacent safes
+      await core.token.updateTransferSteps(from, to, value, hops);
```

Comment:
shouldn't we have try catch here? Otherwise this query goes silent once there is error from it?

Reply:
Yes I thought about it.. but wasn't sure how to do it in a good why, what to do if it fails?

Reply:
I would log the error. We might think about notification as well so it is a clear point where the actual issue is once it happens..

Reply:
I do not see where we catch error from `core.token.updateTransferSteps`

Reply:
I tagged you in a comment with the code